include/mxnet/lib_api.h

```diff
@@ -18,33 +18,756 @@
  */
 
 /*!
- * Copyright (c) 2015 by Contributors
+ * Copyright (c) 2019 by Contributors
  * \file lib_api.h
  * \brief APIs to interact with libraries
+ * This API specifies function prototypes to
+ * register custom ops for library authors
  */
+
 #ifndef MXNET_LIB_API_H_
 #define MXNET_LIB_API_H_
 
+#include <stdint.h>
+#include <stdlib.h>
+#include <vector>
+#include <map>
+#include <string>
+#include <iostream>
+#include <utility>
+
+#define MX_LIBRARY_VERSION 1
+
+/*!
+ * \brief External Tensor data types
+ */
+enum MXDType {
+  kFloat32 = 0,
+  kFloat64 = 1,
+  kFloat16 = 2,
+  kUint8 = 3,
+  kInt32 = 4,
+  kInt8  = 5,
+  kInt64 = 6,
+};
+
+enum MXReturnValue {
+  MX_FAIL = 0,
+  MX_SUCCESS = 1,
+};
+
+/*!
+ * \brief External Tensor data structure
+ */
+struct MXTensor {
+  MXTensor() : data(NULL) {}
+
+  MXTensor(void *data, const std::vector<int64_t> &shape, MXDType dtype)
+  : data(data), shape(shape), dtype(dtype) {}
+
+  /*! \brief helper function to cast data pointer */
+  template<typename data_type>
+  inline data_type* getData() {
```

Comment:
I suggest to check data type in the getData function.
For the name of functions, getData -> data, getDataSize -> size, getDataShape -> shape may be better.

Reply:
I think it should be user's job to figure out what data type to cast given tensor dtype info. The mshadow macro or similar mechanism might be too complex for the custom op, especially it can be done by a few lines of hard-coded types. What do you think?

Reply:
Agree : )